src/c++/perf_analyzer/client_backend/triton_c_api/shared_memory_manager.cc

```diff
@@ -0,0 +1,184 @@
+  std::lock_guard<std::mutex> lock(mu_);
+
+  // If name is already in shared_memory_map_ then return error saying already
+  // registered
+  if (shared_memory_map_.find(name) != shared_memory_map_.end()) {
+    return Error(
+        std::string("shared memory region '" + name + "' already in manager"));
+  }
+
+  shared_memory_map_.insert(std::make_pair(
+      name, std::unique_ptr<MemoryInfo>(new MemoryInfo(
+                name, 0 /* offset */, byte_size, dev_ptr,
+                TRITONSERVER_MEMORY_GPU, device_id))));
+  return Error::Success;
+}
+#endif  // TRITON_ENABLE_GPU
+
+Error
+SharedMemoryManager::RegisterSystemMemory(
+    const std::string& name, void* ptr, const size_t byte_size)
+{
+  // Serialize all operations that write/read current shared memory regions
+  std::lock_guard<std::mutex> lock(mu_);
+
+  // If name is already in shared_memory_map_ then return error saying already
+  // registered
+  if (shared_memory_map_.find(name) != shared_memory_map_.end()) {
+    return Error(
+        std::string("shared memory region '" + name + "' already in manager"));
+  }
+
+  shared_memory_map_.insert(std::make_pair(
+      name, std::unique_ptr<MemoryInfo>(new MemoryInfo(
+                name, 0 /* offset */, byte_size, ptr, TRITONSERVER_MEMORY_CPU,
+                0 /* device id */))));
+
+  return Error::Success;
+}
+
+Error
+SharedMemoryManager::GetMemoryInfo(
+    const std::string& name, size_t offset, void** shm_mapped_addr,
+    TRITONSERVER_MemoryType* memory_type, int64_t* device_id)
+{
+  // protect shared_memory_map_ from concurrent access
+  std::lock_guard<std::mutex> lock(mu_);
+
+  auto it = shared_memory_map_.find(name);
+  if (it == shared_memory_map_.end()) {
+    return Error(
+        std::string("Unable to find shared memory region: '" + name + "'"));
+  }
+  if (it->second->kind_ == TRITONSERVER_MEMORY_CPU) {
+    *shm_mapped_addr =
+        (void*)((uint8_t*)it->second->mapped_addr_ + it->second->offset_ + offset);
+  } else {
+    *shm_mapped_addr = (void*)((uint8_t*)it->second->mapped_addr_ + offset);
+  }
+
+  *memory_type = it->second->kind_;
+  *device_id = it->second->device_id_;
+
+  return Error::Success;
+}
+
+
+Error
+SharedMemoryManager::Unregister(
+    const std::string& name, TRITONSERVER_MemoryType memory_type)
+{
+  // Serialize all operations that write/read current shared memory regions
+  std::lock_guard<std::mutex> lock(mu_);
+
+  return UnregisterHelper(name, memory_type);
+}
+
+Error
+SharedMemoryManager::UnregisterAll(TRITONSERVER_MemoryType memory_type)
+{
+  std::lock_guard<std::mutex> lock(mu_);
+  std::string error_message = "Failed to unregister the following ";
+  std::vector<std::string> unregister_fails;
+  if (memory_type == TRITONSERVER_MEMORY_CPU) {
+    // Serialize all operations that write/read current shared memory regions
+    error_message += "system shared memory regions: ";
+    for (auto it = shared_memory_map_.cbegin(), next_it = it;
```

Comment:
Why not use a range-based for loop for the map?

https://en.cppreference.com/w/cpp/language/range-for

https://stackoverflow.com/a/26282004